sda/protocol.go

```diff
@@ -100,3 +101,99 @@ func ProtocolNameToID(name string) ProtocolID {
+	Wrap(msg interface{}, info *OverlayMessage) (interface{}, error)
+	// Unwrap takes the message coming from the network and must returns the
+	// inner message that is going to be dispatched to the ProtocolInstance, the
+	// OverlayMessage needed by the Overlay to function correctly and then any
+	// error that might have occured.
+	Unwrap(msg interface{}) (interface{}, *OverlayMessage, error)
+	// PacketType returns the packet type ID that this Protocol expects from the
+	// network. This is needed in order for the Overlay to receive those
+	// messages and dispatch them to the correct ProtocolIO.
+	PacketType() network.PacketTypeID
+}
+
+// NewProtocolIO is a function typedef to instantiate a new ProtocolIO
+type NewProtocolIO func() ProtocolIO
+
+type protocolIOFactoryStruct struct {
+	factories map[string]NewProtocolIO
+}
+
+var protocolIOFactory = protocolIOFactoryStruct{
+	factories: make(map[string]NewProtocolIO),
+}
+
+// RegisterProtocolIO takes a name and and NewProtocolIO and save both fields.
+// When a Conode is instantiated, all ProtocolIO will be generated and stored
+// for this Conode.
+func RegisterProtocolIO(name string, n NewProtocolIO) {
+	_, present := protocolIOFactory.factories[name]
+	if present {
+		log.Error("protocolIOStore already registered a ProtocolIO at this name", name)
+		return
+	}
+	protocolIOFactory.factories[name] = n
+}
+
+// protocolIOStore contains all created ProtocolIO and is generally used by the
+// Overlay. It contains the default ProtocolIO used by the Overlay in order to
+// still function properly in case the old wire-format protocol is used.
+type protocolIOStore struct {
+	sync.Mutex
+	protos []ProtocolIO
+	names  map[string]int
+	types  map[network.PacketTypeID]int
+	// the one that gets used in case no ProtocolIO is defined
+	defaultIO ProtocolIO
+}
+
+func (p *protocolIOStore) getByName(name string) ProtocolIO {
+	p.Lock()
+	defer p.Unlock()
+	idx, ok := p.names[name]
+	if !ok || idx >= len(p.protos) || p.protos[idx] == nil {
+		return p.defaultIO
+	}
+	return p.protos[idx]
+}
+
+func (p *protocolIOStore) getByPacketType(t network.PacketTypeID) ProtocolIO {
+	p.Lock()
+	defer p.Unlock()
+	idx, ok := p.types[t]
+	if !ok || idx >= len(p.protos) || p.protos[idx] == nil {
+		return p.defaultIO
+	}
+	return p.protos[idx]
+}
+
+func newProtocolIOStore(disp network.Dispatcher, proc network.Processor) *protocolIOStore {
+	pstore := &protocolIOStore{
+		names:     make(map[string]int),
+		types:     make(map[network.PacketTypeID]int),
+		defaultIO: new(defaultProtoIO),
+	}
+	for name, newIO := range protocolIOFactory.factories {
+		io := newIO()
+		pstore.protos = append(pstore.protos, io)
+		pstore.names[name] = len(pstore.protos) - 1
+		pstore.types[io.PacketType()] = len(pstore.protos) - 1
+		disp.RegisterProcessor(proc, io.PacketType())
+		log.Lvl2("Instantiating ProtocolIO", name, "at position", len(pstore.protos))
+	}
+	// also add the default one
+	return pstore
+}
```

Comment:
And this would also get more easy to write if we don't use the `names` and `types` map.